designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,862 @@
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default [
+
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        ignores: ["**/*.test.js"]
+        extends: [config1, config2],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+The benefit of this approach is that everything is still controlled inside of ESLint itself. There's nothing additional to learn as a user aside from using the `extends` key. 
+
+However, there were several identified problems with this approach:
+
+1. **Older ESLint versions locked out.** Users would have to upgrade to the latest ESLint version to use `extends`, meaning that folks who were still on ESLint v8.x or earlier v9.x would not be able to use this functionality at all.
+1. **Ecosystem compatibility.** There was a danger that plugins could export configs with `extends`, which users in earlier ESLint versions would attempt to use and get an error. Because support for `extends` would be tied to a specific ESLint version, that made it more difficult to create a config that would work in all ESLint versions that support flat config.
+1. **Config Inspector compatibility.** The original plan to implement the functionality in `FlatConfigArray` posed a problem because `FlatConfigArray` is not exported from the `eslint` package. We would either have to export `FlatConfigArray` or implement the functionality in `ConfigArray`. However, `ConfigArray`, being a generic base class, has no concept of plugins, so we'd either have to introduce that concept or not implemented name config extends. There were a lot of different tradeoffs to consider here.
+
+### An `extend()` function
+
+As [mentioned by Kirk Waiblinger](https://github.com/eslint/rfcs/pull/126/files#r1860258048), we could also do a smaller version of `defineConfig()` approach by just defining an `extend()` function that users could use whenever they want to extend a config, such as:
+
+```js
+import { extend } from "eslint";
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+  extend(
+    {
+      files: ["**/src/*.js"],
+    },
+    [js.configs.recommended, example.configs.recommended],
+  ),
+  {
+    files: ["**/*.cjs"],
+    languageOptions: {
+        sourceType: "commonjs"
+    }
+  }
+];
+```
+
+This has similar upsides and downsides to `defineConfig`, with the additional downside that this moves the config further away from the JSON-like visual that makes it easier to scan config files.
+
+## Open Questions
+
+1. **Is `@eslint/config` the right package name?** We could name it `@eslint/define-config`, but that seems like a name that implies `defineConfig()` will be the only export for that package. Using `@eslint/config` seems like it would give us the flexibility to add different kinds of helpers in the future without needing to create a separate package for each. I'm still unsure if this is the correct choice.
```

Comment:
I agree the package should have a more generic name to allow us to add more helpers if needed in the future. but `@eslint/config` sounds like an integral part of ESLint rather than a collection of helpers that could be used directly by end users, so maybe `@eslint/config-utils` or `@eslint/config-helpers`?